Fixes DELETE col[i] support in RULE and COST

craigtaverner opened the pull request:
- COST was missing support for this (in ronja-writes)
- RULE had a new bug introduced during eagerness refactoring (effects)

craigtaverner committed:
Fixes DELETE col[i] support in RULE and COST

* COST was missing support for this (in ronja-writes)
* RULE had a new bug introduced during eagerness refactoring (effects)

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/mutation/DeleteEntityAction.scala:
```diff
@@ -72,10 +72,15 @@ case class DeleteEntityAction(elementToDelete: Expression, forced: Boolean)
-      case _: RelationshipType => Effects(DeletesRelationship)
-      case _: PathType         => Effects(DeletesNode, DeletesRelationship)
+    case i: Variable => effectsFromCypherType(symbols.variables(i.entityName))
+    case ContainerIndex(i: Variable, _) => symbols.variables(i.entityName) match {
```
What about a collection of a collection of nodes (nested)? Is that not supported?

Mats-SX replied:
Just tested it; nested collection is supported, so the `innerType` here can also be `ContainerIndex`. This needs to be recursive.

Mats-SX commented:
Deleting using map lookup should also work (it does in 2.3). That includes nesting as well, ie queries on this form:

```
MATCH (n) WITH {p1: {p2: collect(collect(n))}} AS map DELETE map.p1.p2[0][0]
```

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala:
```diff
@@ -193,7 +197,7 @@ case class UpdateGraph(mutatingPatterns: Seq[MutatingPattern] = Seq.empty) {
    */
   def deleteOverlap(qg: QueryGraph): Boolean = {
-    val identifiersToRead = qg.patternNodes ++ qg.patternRelationships.map(_.name)
+    val identifiersToRead = qg.patternNodes ++ qg.patternRelationships.map(_.name) ++ qg.argumentIds
```
@henriknyman This is what you suggested when we were looking into the `UNWIND` cases -- perhaps it is the correct thing after all.

Mats-SX committed:
Rework how we deal with `DELETE` in the runtime

There's just one `DeletePipe` now, because the code was so similar between the
three cases of entities to delete: nodes, relationships and paths. Also added
support in the cost planner to delete from maps and nested map/collections.

Mats-SX commented:
Rebased and updated licenses.
